src/client.ts

```diff
@@ -0,0 +1,583 @@
+  ) {
+    const u = new URL(url);
+    opts = Object.assign(opts || {}, {
+      hostname: u.hostname,
+      port: u.port,
+      path: u.pathname + u.search,
+      method: method,
+      headers: {},
+    });
+
+    const authToken = await this.#auth.getAccessToken();
+    if (!authToken) {
+      throw new Error(`Failed to get auth token for ${method} ${url}`);
+    }
+
+    opts.headers = Object.assign(
+      {
+        'Authorization': `Bearer ${authToken}`,
+        'Accept': 'application/json',
+        'Content-Type': 'application/json',
+      },
+      opts.headers,
+    );
+
+    try {
+      const resp = await CloudFunctionsClient.request(opts, data);
+      return JSON.parse(resp);
+    } catch (err) {
+      throw new Error(`Failed to ${method} ${url}: ${errorMessage(err)}`);
+    }
+  }
+
+  /**
+   * pollOperation polls the operation, calling pollFn on each attempt.
+   *
+   * @param name Name of the operation, of the format `operations/{name}`.
+   * @param opts Options for polling
+   */
+  async #pollOperation(
+    name: string,
+    opts: PollOperationOptions = { interval: 5, retries: 60 },
+  ): Promise<Operation> {
+    const intervalMs: number = +opts.interval * 1000;
+
+    for (let i = 0; i < opts.retries; i++) {
+      // If a poll function was given, call it.
+      if (opts.onPoll) opts.onPoll();
+
+      const resp = await this.getOperation(name);
+      if (resp.error) {
+        throw new Error(`Operation failed: ${resp.error.message}`);
+      }
+      if (resp.done) {
+        return resp;
+      }
+
+      await new Promise((resolve) => setTimeout(resolve, intervalMs));
+    }
+
+    throw new Error(`Operation timed out`);
+  }
+
+  /**
+   * getOperation fetches the operation by name.
+   *
+   * @param name Name of the operation, of the format `operations/{name}`.
+   */
+  async getOperation(name: string): Promise<Operation> {
+    if (!name.startsWith('operations/')) {
+      name = `operations/${name}`;
+    }
+
+    const u = `${this.#baseURL}/${name}`;
+    const resp: Operation = await this.#request('GET', u);
+    return resp;
+  }
+
+  /**
+   * create creates a new Cloud Function.
+   *
+   * @param cf Cloud Function to deploy.
+   */
+  async create(
+    cf: CloudFunction,
+    opts?: CreateOptions,
+  ): Promise<CloudFunctionResponse> {
+    const timeout = opts?.timeout || defaultTimeout;
+
+    const resourceName = this.fullResourceName(cf.name);
+    cf.name = resourceName;
+
+    const parent = this.parentFromName(resourceName);
+    const u = `${this.#baseURL}/${parent}/functions`;
+    const body = JSON.stringify(cf);
+    const resp: Operation = await this.#request('POST', u, {}, body);
+    const op = await this.#pollOperation(resp.name, {
+      interval: 5,
+      retries: timeout / 5,
+      onPoll: opts?.onPoll,
+    });
+
+    if (!op.response) {
+      throw new Error(`create operation result did not include function`);
```

Comment:
I think it would be a GCP bug. I need the assertion to make TypeScript happy though, since the `response` is `response?`, since it's not included if there's an error. The error condition is handled by `pollOperation`, but I need the type assertion here.

Reply:
Sgtm I doubt users will hit this. Should we modify the message to `create operation result did not return a response`? I think `include a function ` does not make sense unless familiar with API